cmd/openshift-install/create.go

```diff
@@ -655,3 +717,177 @@ func checkIfAgentCommand(assetStore asset.Store) {
+				}
+				logrus.Debugf("Cluster Operator %s is Progressing=%s LastTransitionTime=%v DurationSinceTransition=%.fs Reason=%s Message=%s", name, progressing.Status, progressing.LastTransitionTime.Time, time.Since(progressing.LastTransitionTime.Time).Seconds(), progressing.Reason, progressing.Message)
+			}
+			return false, nil
+		},
+	)
+	return pStatus, err
+}
+
+func wrapWithReplay(w watch.Interface) watch.Interface {
+	fw := &replayingWatch{
+		incoming: w,
+		result:   make(chan watch.Event),
+		closed:   make(chan struct{}),
+	}
+	go fw.watchIncoming()
+	go fw.resendLast()
+	return fw
+}
+
+type replayingWatch struct {
+	incoming watch.Interface
+	result   chan watch.Event
+	closed   chan struct{}
+
+	lastLock sync.Mutex
+	last     watch.Event
+}
+
+func (r *replayingWatch) ResultChan() <-chan watch.Event {
+	return r.result
+}
+
+func (r *replayingWatch) Stop() {
+	r.incoming.Stop()
+}
+
+func (r *replayingWatch) watchIncoming() {
+	defer close(r.result)
+	defer close(r.closed)
+	for event := range r.incoming.ResultChan() {
+		func() {
+			r.lastLock.Lock()
+			defer r.lastLock.Unlock()
+
+			r.result <- event
+			r.last = copyWatchEvent(event)
+		}()
+	}
+}
+func copyWatchEvent(event watch.Event) watch.Event {
+	return watch.Event{
+		Type:   event.Type,
+		Object: event.Object.DeepCopyObject(),
+	}
+}
+
+func (r *replayingWatch) resendLast() {
+	var emptyEvent watch.Event
+	err := wait.PollUntilContextCancel(context.TODO(), time.Second, false, func(ctx context.Context) (bool, error) {
+		select {
+		case <-r.closed:
+			return true, nil
+		default:
+		}
+		func() {
+			r.lastLock.Lock()
+			defer r.lastLock.Unlock()
+
+			if r.last != emptyEvent {
+				r.result <- copyWatchEvent(r.last)
+			}
+		}()
+		return false, nil
+	})
+	if err != nil {
+		logrus.Debugf("Watcher polling error: %v", err)
+	}
+}
+
+func replayingListWatcher(in cache.ListerWatcher) cache.ListerWatcher {
+	return &cache.ListWatch{
+		ListFunc: in.List,
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			w, err := in.Watch(options)
+			if err != nil {
+				return w, err
+			}
+			return wrapWithReplay(w), nil
+		},
+		DisableChunking: true,
+	}
+}
+
+// coStabilityChecker returns a closure which references a shared error variable. err
+// tracks whether any operator has had a stability error. The closure function will
+// return an error if any operator has had an instability error, even if the operator
+// currently being checked is stable.
+func coStabilityChecker() func(string, *configv1.ClusterOperatorStatusCondition, error) error {
+	var err error
+
+	return func(name string, status *configv1.ClusterOperatorStatusCondition, statusErr error) error {
+		if statusErr == nil {
+			return err
+		}
+		if !wait.Interrupted(statusErr) {
+			logrus.Errorf("Error checking cluster operator %s Progressing status: %q", name, statusErr)
+			logrus.Exit(exitCodeOperatorStabilityFailed)
+			err = errors.New("cluster operators are not stable")
```

Comment:
Is this reachable code?  It looks like [`logrus.Exit(...)`][1] is fatal, but maybe I'm missing some context?

[1]: https://pkg.go.dev/github.com/sirupsen/logrus#Exit

Reply:
@patrickdillon do you remember if `logrus.Exit` is just an artifact of a trial phase? It does look like we want to set and return the error, not exit the install immediately.

Reply:
`logrus.Exit(exitCode...)` looks like existing precedent to me.  But yes, I would expect to be returnning `err` here, and then using `logrus.Exit(exitCodeOperatorStabilityFailed)` right after the `wg.Wait()` if `err` wasn't `nil`.

Reply:
The pattern trevor suggested makes sense. I can't remember the motivation for originally writing this code. Perhaps the idea was to fail quickly if there were problems checking the status, but it definitely looks like it could be improved

Reply:
Done